### PR TITLE
Pass env_id to replay buffer methods to correctly support batch training

### PR DESCRIPTION
**Pull request by muupan**
~Merge #443 before this PR.~

Current replay buffers with num_steps > 1 or episodic are not correct in batch training because they cannot know which env a given transition came from.

This PR adds the `env_id` argument to two methods of replay buffers: `append` and `stop_current_episode`. From `env_id` replay buffers can know which env a given transition came from and which env's episode is stopped.

TODO:
- [x] check how it affects scores with n-step return and batch training

### CODE REVIEW

**Comment by muupan** [comment]
I checked the effect of the bug fixed by this PR. I added the `--n-step-return` option to `examples/ale/train_dqn_batch_ale.py` and run it with `--n-step-return 1` and `--n-step-return 3`, using 3 different random seeds:
```
python3 examples/ale/train_dqn_batch_ale.py --num-envs 8 --n-step-return 1 --steps 10000000 --env SpaceInvadersNoFrameskip-v4
```
```
python3 examples/ale/train_dqn_batch_ale.py --num-envs 8 --n-step-return 3 --steps 10000000 --env SpaceInvadersNoFrameskip-v4
```

Before this PR, `--n-step-return 3` completely failed. After this PR, `--n-step-return 3` learned faster than `--n-step-return 1` as expected.

image

File: chainerrl/replay_buffer.py [hunk 1]
```diff
@@ -135,35 +143,44 @@ def __init__(self, capacity=None, num_steps=1):
...
+            state=state,
+            action=action,
+            reward=reward,
+            next_state=next_state,
+            next_action=next_action,
+            is_state_terminal=is_state_terminal,
+            **kwargs
+        )
+        last_n_transitions.append(experience)
         if is_state_terminal:
```
Comment (prabhatnagarajan) [inline chainerrl/replay_buffer.py]: Why do we remove the if statement that checks if the transition is terminal?

File: tests/test_replay_buffer.py [hunk 2]
```diff
@@ -487,6 +487,101 @@ def test_append_and_sample(self):
...
+        # 9 transitions for env_id=2
+        for i in range(9):
+            trans1 = dict(state=0, action=1, reward=2, next_state=3,
+                          next_action=4, is_state_terminal=False)
+            rbuf.append(env_id=2, **trans1)
+
+        # It should have:
+        #   - 4 transitions from env_id=1
+        #   - 5 transitions from env_id=2
+        self.assertEqual(len(rbuf), 9)
```
Comment (prabhatnagarajan) [inline tests/test_replay_buffer.py]: Shouldn't we have 3 transitions? One from `env_id=1` and five from `env_id=2`?

File: chainerrl/replay_buffer.py (hunk 1 above)
Comment (muupan) [reply chainerrl/replay_buffer.py]: i think it is not removed

File: tests/test_replay_buffer.py (hunk 2 above)
Comment (muupan) [reply tests/test_replay_buffer.py]: Four form `env_id=1`: (s_0, s_4), (s_1, s_4), (s_2, s_4), and (s_3, s_4).

Comment (prabhatnagarajan) [reply tests/test_replay_buffer.py]: Got it.

**Pull request merged**